File: ts/test/polling.test.ts
import { describe, expect, test } from "bun:test";
import { createPolling } from "../src/core/polling.ts";
import { createAppState } from "../src/core/state.ts";
import { defaultSettings } from "../src/core/settings.ts";
import { dtFromParts, serdePrettyJson } from "../src/core/json.ts";
import { quotaResultToSerde, type QuotaResult } from "../src/core/quota.ts";

const AT = dtFromParts(1_893_456_000_000, 0);

const result = (error: string | null, label: string): QuotaResult => ({
  fiveHour: error === null ? { percent: 21, resetAt: null } : null,
  week: error === null ? { percent: 18, resetAt: null } : null,
  extra: error === null ? { state: "Ready", balanceCents: 1n, monthlyEnabled: false, monthlyUsedCents: null, monthlyLimitCents: null } : null,
  fetchedAt: AT,
  error,
});

interface Harness {
  scheduled: { ms: number; run: () => void }[];
  published: QuotaResult[];
  fire: () => Promise<void>;
  polling: ReturnType<typeof createPolling>;
  state: ReturnType<typeof createAppState>;
  queue: (r: QuotaResult) => void;
}

const harness = (first: QuotaResult = result(null, "a")): Harness => {
  const scheduled: { ms: number; run: () => void }[] = [];
  const published: QuotaResult[] = [];
  const queue: QuotaResult[] = [first];
  const state = createAppState(defaultSettings(), "light");
  const polling = createPolling({
    state,
    fetchQuota: async () => queue.shift() ?? result(null, "x"),
    publish: (r) => void published.push(r),
    setTimer: (ms, run) => {
      scheduled.push({ ms, run });
      return () => {
        const at = scheduled.findIndex((s) => s.run === run);
        if (at >= 0) scheduled.splice(at, 1);
      };
    },
  });
  return {
    scheduled,
    published,
    polling,
    state,
    queue: (r) => void queue.push(r),
    fire: async () => {
      const next = scheduled.shift();
      // A macrotask flush: the serialized refresh chain takes a few more
      // microtask hops than a bare await can drain before the tick re-arms.
      if (next) await next.run();
      await new Promise((r) => setTimeout(r, 0));
    },
  };
};

const lastDelay = (h: Harness): number => h.scheduled[h.scheduled.length - 1]!.ms;

describe("polling schedule (SPEC 16.5)", () => {
  test("the first refresh fires two seconds after start", () => {
    const h = harness();
    h.polling.start();
    expect(h.scheduled).toHaveLength(1);
    expect(h.scheduled[0]!.ms).toBe(2000);
    expect(h.published).toHaveLength(0);
  });

  test("a success re-arms at the configured period", async () => {
    const h = harness();
    h.polling.start();
    await h.fire();
    expect(lastDelay(h)).toBe(5 * 60_000);
    expect(h.published).toHaveLength(1);
    expect(h.state.lastQuota).not.toBeNull();
  });

  test("a failure re-arms after 30 seconds", async () => {
    const h = harness(result("HttpRequestException", "bad"));
    h.polling.start();
    await h.fire();
    expect(lastDelay(h)).toBe(30_000);
  });

  test("the period is read when the tick is scheduled, so a save takes effect", async () => {
    const h = harness();
    h.polling.start();
    await h.fire();
    h.state.settings.refreshMinutes = 30;
    await h.fire();
    expect(lastDelay(h)).toBe(30 * 60_000);
  });

  test("refresh_minutes below one is floored to one minute", async () => {
    const h = harness();
    h.state.settings.refreshMinutes = 0;
    h.polling.start();
    await h.fire();
    expect(lastDelay(h)).toBe(60_000);
  });

  test("an above-ceiling period clamps to the setTimeout maximum", async () => {
    const h = harness();
    h.state.settings.refreshMinutes = Number.MAX_SAFE_INTEGER;
    h.polling.start();
    await h.fire();
    // Unclamped, 2^53 * 60_000 overflows the int32 setTimeout and would
    // reschedule at 1 ms — a hot loop hammering the API.
    expect(lastDelay(h)).toBe(2_147_483_647);
  });
});

describe("keep-last-good (SPEC 16.5 step 2)", () => {
  test("a failed refresh keeps the previous data and the new stamp", async () => {
    const h = harness();
    h.polling.start();
    await h.fire(); // good
    h.queue(result("HttpRequestException", "bad"));
    await h.fire();
    const last = h.published[h.published.length - 1]!;
    expect(last.error).toBe("HttpRequestException");
    expect(last.fiveHour?.percent).toBe(21); // filled from last-good
    expect(last.fetchedAt).toBe(AT);
  });

  test("the first failure with nothing cached stays empty", async () => {
    const h = harness(result("no-token", "none"));
    h.polling.start();
    await h.fire();
    const last = h.published[0]!;
    expect(last.fiveHour).toBeNull();
    expect(last.error).toBe("no-token");
  });
});

describe("reschedule and manual refresh", () => {
  test("reschedule forces the 2 second delay and drains a stale hint", async () => {
    const h = harness(result("HttpRequestException", "bad"));
    h.polling.start();
    await h.fire(); // failure => 30s pending
    expect(lastDelay(h)).toBe(30_000);
    h.polling.reschedule();
    expect(lastDelay(h)).toBe(2000);
    await h.fire();
    expect(h.published).toHaveLength(2);
  });

  test("a manual refresh moves the next tick to the same rule", async () => {
    const h = harness();
    h.polling.start();
    expect(lastDelay(h)).toBe(2000);
    await h.polling.safeRefresh();
    expect(lastDelay(h)).toBe(5 * 60_000);
    expect(h.published).toHaveLength(1);
  });

  test("a manual refresh after a failure arms the fast retry", async () => {
    const h = harness(result("TaskCanceledException", "slow"));
    h.polling.start();
    await h.polling.safeRefresh();
    expect(lastDelay(h)).toBe(30_000);
  });

  test("stop cancels the pending tick", async () => {
    const h = harness();
    h.polling.start();
    h.polling.stop();
    await h.fire();
    expect(h.published).toHaveLength(0);
  });
});

describe("arm epoch and in-flight serialization (SPEC 16.5)", () => {
  /** A harness whose fetches stay in flight until the test resolves them. */
  const blockingHarness = () => {
    const scheduled: { ms: number; run: () => void }[] = [];
    const published: QuotaResult[] = [];
    const resolvers: ((r: QuotaResult) => void)[] = [];
    const state = createAppState(defaultSettings(), "light");
    const polling = createPolling({
      state,
      fetchQuota: () => new Promise<QuotaResult>((res) => resolvers.push(res)),
      publish: (r) => void published.push(r),
      setTimer: (ms, run) => {
        scheduled.push({ ms, run });
        return () => {
          const at = scheduled.findIndex((s) => s.run === run);
          if (at >= 0) scheduled.splice(at, 1);
        };
      },
    });
    const flush = () => new Promise((r) => setTimeout(r, 0));
    return { scheduled, published, resolvers, state, polling, flush };
  };

  test("an in-flight tick must not cancel a reschedule armed while it awaited", async () => {
    const h = blockingHarness();
    h.polling.start();
    const first = h.scheduled.shift()!;
    expect(first.ms).toBe(2000);
    first.run(); // the tick's fetch is now in flight
    // Settings saved mid-fetch: reschedule arms a fresh 2s tick.
    h.polling.reschedule();
    expect(lastDelayMs(h.scheduled)).toBe(2000);
    // The in-flight tick completes; without the epoch guard it would re-arm
    // 30s/period over the reschedule's 2s tick.
    h.resolvers[0]!(result(null, "late"));
    await h.flush();
    expect(lastDelayMs(h.scheduled)).toBe(2000);
  });

  test("an in-flight manual refresh must not re-arm over a reschedule", async () => {
    const h = blockingHarness();
    h.polling.start();
    const manual = h.polling.safeRefresh();
    h.polling.reschedule();
    expect(lastDelayMs(h.scheduled)).toBe(2000);
    h.resolvers[0]!(result(null, "late"));
    await manual;
    expect(lastDelayMs(h.scheduled)).toBe(2000);
  });

  test("a manual refresh queues behind an in-flight tick instead of racing it", async () => {
    const h = blockingHarness();
    h.polling.start();
    h.scheduled.shift()!.run(); // tick fetch #1 in flight
    const manual = h.polling.safeRefresh();
    // No second fetch while the first is still in flight.
    expect(h.resolvers).toHaveLength(1);
    const a = result(null, "a");
    a.fiveHour!.percent = 1;
    h.resolvers[0]!(a);
    await h.flush(); // the queued manual fetch starts now
    expect(h.resolvers).toHaveLength(2);
    const b = result(null, "b");
    b.fiveHour!.percent = 2;
    h.resolvers[1]!(b);
    await manual;
    // Publish order follows fetch order — the older result can never land last.
    expect(h.published.map((r) => r.fiveHour!.percent)).toEqual([1, 2]);
    expect(h.state.lastQuota!.fiveHour!.percent).toBe(2);
  });
});

const lastDelayMs = (scheduled: { ms: number }[]): number => scheduled[scheduled.length - 1]!.ms;

test("the published shape is what --test-fetch prints", async () => {
  const h = harness();
  h.polling.start();
  await h.fire();
  expect(serdePrettyJson(quotaResultToSerde(h.published[0]!))).toContain('"percent": 21.0');
});

File: go/internal/core/polling_test.go
package core

import (
	"math"
	"strings"
	"sync"
	"testing"
	"time"
)

// fakeSetTimer records scheduled ticks; the test fires them manually.
type scheduledTick struct {
	id  int
	ms  int64
	run func()
}

type fakeTimer struct {
	mu     sync.Mutex
	ticks  []*scheduledTick
	nextID int
}

func (f *fakeTimer) setTimer(ms int64, run func()) func() {
	f.mu.Lock()
	f.nextID++
	tick := &scheduledTick{id: f.nextID, ms: ms, run: run}
	f.ticks = append(f.ticks, tick)
	f.mu.Unlock()
	return func() {
		f.mu.Lock()
		defer f.mu.Unlock()
		for i, t := range f.ticks {
			if t.id == tick.id {
				f.ticks = append(f.ticks[:i], f.ticks[i+1:]...)
				return
			}
		}
	}
}

func (f *fakeTimer) lastDelay() int64 {
	f.mu.Lock()
	defer f.mu.Unlock()
	if len(f.ticks) == 0 {
		return -1
	}
	return f.ticks[len(f.ticks)-1].ms
}

func (f *fakeTimer) shift() *scheduledTick {
	f.mu.Lock()
	defer f.mu.Unlock()
	if len(f.ticks) == 0 {
		return nil
	}
	t := f.ticks[0]
	f.ticks = f.ticks[1:]
	return t
}

func (f *fakeTimer) pending() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return len(f.ticks)
}

type pollingHarness struct {
	timer     *fakeTimer
	published []QuotaResult
	polling   *Polling
	state     *AppState
	queue     []QuotaResult
}

func goodResult(label string) QuotaResult {
	return QuotaResult{
		FiveHour:  &QuotaSegment{Percent: 21},
		Week:      &QuotaSegment{Percent: 18},
		Extra:     &ExtraInfo{State: ExtraReady, BalanceCents: ptrI64(1)},
		FetchedAt: atZero,
	}
}

func failedResult(kind string) QuotaResult {
	return QuotaResult{FetchedAt: atZero, Error: &kind}
}

func newHarness(first QuotaResult) *pollingHarness {
	h := &pollingHarness{
		timer: &fakeTimer{},
		queue: []QuotaResult{first},
		state: NewAppState(DefaultSettings(), "light"),
	}
	h.polling = NewPolling(PollingDeps{
		State: h.state,
		FetchQuota: func() QuotaResult {
			if len(h.queue) == 0 {
				return goodResult("x")
			}
			r := h.queue[0]
			h.queue = h.queue[1:]
			return r
		},
		Publish:  func(r QuotaResult) { h.published = append(h.published, r) },
		SetTimer: h.timer.setTimer,
	})
	return h
}

func (h *pollingHarness) fire() {
	if tick := h.timer.shift(); tick != nil {
		tick.run()
	}
}

func TestPollingFirstRefreshFiresTwoSecondsAfterStart(t *testing.T) {
	h := newHarness(goodResult("a"))
	h.polling.Start()
	if h.timer.pending() != 1 || h.timer.lastDelay() != 2000 {
		t.Errorf("start arms one 2 s tick: pending=%d delay=%d", h.timer.pending(), h.timer.lastDelay())
	}
	if len(h.published) != 0 {
		t.Error("no refresh runs before the tick fires")
	}
}

func TestPollingSuccessRearmsAtPeriod(t *testing.T) {
	h := newHarness(goodResult("a"))
	h.polling.Start()
	h.fire()
	if h.timer.lastDelay() != 5*60_000 {
		t.Errorf("success re-arms at the period, got %d", h.timer.lastDelay())
	}
	if len(h.published) != 1 || h.state.LastQuota() == nil {
		t.Errorf("published=%d lastQuota=%+v", len(h.published), h.state.LastQuota())
	}
}

func TestPollingFailureRearmsAfter30Seconds(t *testing.T) {
	h := newHarness(failedResult("HttpRequestException"))
	h.polling.Start()
	h.fire()
	if h.timer.lastDelay() != 30_000 {
		t.Errorf("failure re-arms at 30 s, got %d", h.timer.lastDelay())
	}
}

func TestPollingPeriodReadAtTickTime(t *testing.T) {
	h := newHarness(goodResult("a"))
	h.polling.Start()
	h.fire()
	{
		s := h.state.Settings()
		s.RefreshMinutes = 30
		h.state.SetSettings(s)
	}
	h.fire()
	if h.timer.lastDelay() != 30*60_000 {
		t.Errorf("a save must take effect on the next tick, got %d", h.timer.lastDelay())
	}
}

func TestPollingRefreshMinutesBelowOneFlooredToOneMinute(t *testing.T) {
	h := newHarness(goodResult("a"))
	{
		s := h.state.Settings()
		s.RefreshMinutes = 0
		h.state.SetSettings(s)
	}
	h.polling.Start()
	h.fire()
	if h.timer.lastDelay() != 60_000 {
		t.Errorf("minutes 0 floors to one minute, got %d", h.timer.lastDelay())
	}
}

// REVIEW-M1 Major A / PLAN-GO §4.4: a huge refreshMinutes must mean "long
// sleep", not an overflow to a short hot loop. Go has no int32 setTimeout
// ceiling, so the correct answer is the saturating maximum, not 2^31-1.
func TestPollingHugeRefreshMinutesNeverShortDelays(t *testing.T) {
	if got := periodMs(math.MaxInt64); got != math.MaxInt64 {
		t.Errorf("periodMs(i64::MAX) = %d, want i64::MAX", got)
	}
	if got := periodMs(0); got != 60_000 {
		t.Errorf("periodMs(0) = %d, want one minute", got)
	}
	if d := msToDuration(math.MaxInt64); d <= 0 {
		t.Errorf("msToDuration must saturate, got %v", d)
	}
	h := newHarness(goodResult("a"))
	{
		s := h.state.Settings()
		s.RefreshMinutes = math.MaxInt64
		h.state.SetSettings(s)
	}
	h.polling.Start()
	h.fire()
	if h.timer.lastDelay() != math.MaxInt64 {
		t.Errorf("huge period saturates, got %d", h.timer.lastDelay())
	}
}

func TestPollingKeepsLastGoodOnFailure(t *testing.T) {
	h := newHarness(goodResult("a"))
	h.polling.Start()
	h.fire() // good
	h.queue = append(h.queue, failedResult("HttpRequestException"))
	h.fire()
	last := h.published[len(h.published)-1]
	if last.ErrorKind() != "HttpRequestException" {
		t.Errorf("error = %q", last.ErrorKind())
	}
	if last.FiveHour == nil || last.FiveHour.Percent != 21 {
		t.Errorf("filled from last-good: %+v", last.FiveHour)
	}
	if last.FetchedAt != atZero {
		t.Errorf("fetchedAt = %+v", last.FetchedAt)
	}
}

func TestPollingFirstFailureStaysEmpty(t *testing.T) {
	h := newHarness(failedResult("no-token"))
	h.polling.Start()
	h.fire()
	if len(h.published) != 1 || h.published[0].FiveHour != nil || h.published[0].ErrorKind() != "no-token" {
		t.Errorf("first failure stays empty: %+v", h.published[0])
	}
}

func TestPollingRescheduleDrainsStaleHint(t *testing.T) {
	h := newHarness(failedResult("HttpRequestException"))
	h.polling.Start()
	h.fire() // failure => 30 s pending
	if h.timer.lastDelay() != 30_000 {
		t.Fatalf("precondition: 30 s pending, got %d", h.timer.lastDelay())
	}
	h.polling.Reschedule()
	if h.timer.lastDelay() != 2000 {
		t.Errorf("reschedule forces 2 s, got %d", h.timer.lastDelay())
	}
	h.fire()
	if len(h.published) != 2 {
		t.Errorf("published=%d, want 2", len(h.published))
	}
}

func TestPollingManualRefreshMovesNextTick(t *testing.T) {
	h := newHarness(goodResult("a"))
	h.polling.Start()
	if h.timer.lastDelay() != 2000 {
		t.Fatalf("precondition 2 s, got %d", h.timer.lastDelay())
	}
	h.polling.SafeRefresh()
	if h.timer.lastDelay() != 5*60_000 {
		t.Errorf("manual refresh moves the tick to the period, got %d", h.timer.lastDelay())
	}
	if len(h.published) != 1 {
		t.Errorf("published=%d, want 1", len(h.published))
	}
}

func TestPollingManualRefreshAfterFailureArmsFastRetry(t *testing.T) {
	h := newHarness(failedResult("TaskCanceledException"))
	h.polling.Start()
	h.polling.SafeRefresh()
	if h.timer.lastDelay() != 30_000 {
		t.Errorf("manual refresh after failure arms 30 s, got %d", h.timer.lastDelay())
	}
}

func TestPollingStopCancelsPendingTick(t *testing.T) {
	h := newHarness(goodResult("a"))
	h.polling.Start()
	h.polling.Stop()
	h.fire()
	if len(h.published) != 0 {
		t.Errorf("stop cancels the pending tick, published=%d", len(h.published))
	}
}

// ---------------------------------------------------------------------------
// arm epoch and in-flight serialization (SPEC 16.5)
// ---------------------------------------------------------------------------

type blockingHarness struct {
	timer     *fakeTimer
	published []QuotaResult
	resolvers []chan QuotaResult
	polling   *Polling
	state     *AppState
	mu        sync.Mutex
	done      []chan struct{}
}

func newBlockingHarness() *blockingHarness {
	b := &blockingHarness{
		timer: &fakeTimer{},
		state: NewAppState(DefaultSettings(), "light"),
	}
	b.polling = NewPolling(PollingDeps{
		State: b.state,
		FetchQuota: func() QuotaResult {
			ch := make(chan QuotaResult, 1)
			b.mu.Lock()
			b.resolvers = append(b.resolvers, ch)
			b.mu.Unlock()
			r := <-ch
			return r
		},
		Publish: func(r QuotaResult) {
			b.mu.Lock()
			b.published = append(b.published, r)
			b.mu.Unlock()
		},
		SetTimer: b.timer.setTimer,
	})
	return b
}

func (b *blockingHarness) resolve(idx int, r QuotaResult) {
	b.mu.Lock()
	ch := b.resolvers[idx]
	b.mu.Unlock()
	ch <- r
}

func (b *blockingHarness) resolverCount() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return len(b.resolvers)
}

func TestInFlightTickMustNotCancelReschedule(t *testing.T) {
	b := newBlockingHarness()
	b.polling.Start()
	first := b.timer.shift()
	if first == nil || first.ms != 2000 {
		t.Fatalf("precondition: one 2 s tick, got %+v", first)
	}
	go first.run() // the tick's fetch is now in flight
	waitFor(t, func() bool { return b.resolverCount() == 1 })

	// Settings saved mid-fetch: reschedule arms a fresh 2 s tick.
	b.polling.Reschedule()
	if b.timer.lastDelay() != 2000 {
		t.Fatalf("reschedule arms 2 s, got %d", b.timer.lastDelay())
	}

	// The in-flight tick completes; without the epoch guard it would re-arm
	// 30 s/period over the reschedule's 2 s tick.
	b.resolve(0, goodResult("late"))
	waitFor(t, func() bool { return b.timer.lastDelay() == 2000 && b.timer.pending() == 1 })
}

func TestInFlightManualRefreshMustNotRearmOverReschedule(t *testing.T) {
	b := newBlockingHarness()
	b.polling.Start()

	done := make(chan QuotaResult, 1)
	go func() { done <- b.polling.SafeRefresh() }()
	waitFor(t, func() bool { return b.resolverCount() == 1 })

	b.polling.Reschedule()
	if b.timer.lastDelay() != 2000 {
		t.Fatalf("reschedule arms 2 s, got %d", b.timer.lastDelay())
	}
	b.resolve(0, goodResult("late"))
	<-done
	waitFor(t, func() bool { return b.timer.lastDelay() == 2000 && b.timer.pending() == 1 })
}

func TestManualRefreshQueuesBehindInFlightTick(t *testing.T) {
	b := newBlockingHarness()
	b.polling.Start()
	tick := b.timer.shift()
	go tick.run()
	waitFor(t, func() bool { return b.resolverCount() == 1 })

	done := make(chan QuotaResult, 1)
	go func() { done <- b.polling.SafeRefresh() }()

	// No second fetch while the first is still in flight.
	if b.resolverCount() != 1 {
		t.Fatalf("manual refresh must queue, resolvers=%d", b.resolverCount())
	}

	a := goodResult("a")
	a.FiveHour = &QuotaSegment{Percent: 1}
	b.resolve(0, a)
	waitFor(t, func() bool { return b.resolverCount() == 2 })

	b2 := goodResult("b")
	b2.FiveHour = &QuotaSegment{Percent: 2}
	b.resolve(1, b2)
	<-done

	// Publish order follows fetch order — the older result can never land last.
	b.mu.Lock()
	defer b.mu.Unlock()
	if len(b.published) != 2 || b.published[0].FiveHour.Percent != 1 || b.published[1].FiveHour.Percent != 2 {
		t.Errorf("publish order: %+v", b.published)
	}
	if b.state.LastQuota() == nil || b.state.LastQuota().FiveHour.Percent != 2 {
		t.Errorf("lastQuota must hold the newest result: %+v", b.state.LastQuota())
	}
}

func TestAppStateDefaults(t *testing.T) {
	s := NewAppState(DefaultSettings(), "dark")
	if s.LastQuota() != nil || s.SkillsCache() != nil || s.LastManualRefreshMs() != nil {
		t.Errorf("fresh state must be empty: %+v", s)
	}
	if s.EffectiveTheme() != "dark" || s.Update().CheckFailed || s.Update().UpdateAvailable {
		t.Errorf("fresh state fields: %+v", s)
	}
	if !strings.Contains(SettingsToJSONText(s.Settings()), `"Theme": "system"`) {
		t.Errorf("settings shape: %s", SettingsToJSONText(s.Settings()))
	}
}

// M1 review Minor 3: between two serialized refreshes the LATER finisher
// decides the next delay, exactly like Rust's retime notify — a tick failure
// (30 s) must not survive a queued manual success (period).
func TestQueuedManualRefreshRetimeWins(t *testing.T) {
	b := newBlockingHarness()
	b.polling.Start()
	tick := b.timer.shift()
	go tick.run()
	waitFor(t, func() bool { return b.resolverCount() == 1 })

	done := make(chan QuotaResult, 1)
	go func() { done <- b.polling.SafeRefresh() }()

	// The tick's fetch fails: its own completion arms the 30 s fast retry,
	// and only then does the queued manual fetch start.
	b.resolve(0, failedResult("HttpRequestException"))
	waitFor(t, func() bool { return b.timer.lastDelay() == 30_000 })
	waitFor(t, func() bool { return b.resolverCount() == 2 })

	// The queued manual refresh succeeds: the period must win.
	b.resolve(1, goodResult("manual"))
	<-done
	waitFor(t, func() bool { return b.timer.lastDelay() == 5*60_000 })
}

// waitFor polls cond with a deadline; a plain sleep would make the epoch
// tests flaky, a busy loop without deadline would hang on regression.
func waitFor(t *testing.T, cond func() bool) {
	t.Helper()
	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) {
		if cond() {
			return
		}
		time.Sleep(500 * time.Microsecond)
	}
	t.Fatal("condition not reached within 2 s")
}
